File: sdks/typescript/src/telemetry.ts
/**
 * Background telemetry shipping.
 *
 * Design goals (from the spec):
 *   - The synchronous hot path is a single bounded-array push.
 *   - A background timer batches events by count (N) or time (T) and POSTs them.
 *   - Graceful degradation: unreachable ingest -> log + drop (after retries).
 *     A 429 backpressure response is retried with backoff, not dropped.
 *   - Bounded buffer: when full, the new event is dropped + a warning logged,
 *     never blocking the caller.
 */

import type { Event } from "./schema.js";

/* eslint-disable no-console */
const warn = (msg: string, ...args: unknown[]): void => {
  console.warn(`costobs: ${msg}`, ...args);
};

export interface Sender {
  /** POST a batch. Resolve with the HTTP status code, or null on transport error. */
  send(events: Array<Record<string, unknown>>): Promise<number | null>;
  close?(): void | Promise<void>;
}

/** Default fetch-based POST sender. */
export class HttpSender implements Sender {
  private readonly url: string;
  private readonly apiKey: string;
  private readonly timeoutMs: number;

  constructor(baseUrl: string, apiKey: string, timeoutMs = 5000) {
    this.url = baseUrl.replace(/\/+$/, "") + "/v1/events";
    this.apiKey = apiKey;
    this.timeoutMs = timeoutMs;
  }

  async send(events: Array<Record<string, unknown>>): Promise<number | null> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const resp = await fetch(this.url, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ events }),
        signal: controller.signal,
      });
      return resp.status;
    } catch {
      // Transport error (network down, abort, DNS, …).
      return null;
    } finally {
      clearTimeout(timer);
    }
  }
}

export interface TelemetryOptions {
  maxQueue?: number;
  batchSize?: number;
  flushIntervalMs?: number;
  maxRetries?: number;
  sender?: Sender;
  timeoutMs?: number;
}

const sleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

/** Bounded in-memory buffer + background flush loop. */
export class TelemetryQueue {
  private readonly buffer: Array<Record<string, unknown>> = [];
  private readonly maxQueue: number;
  private readonly batchSize: number;
  private readonly flushIntervalMs: number;
  private readonly maxRetries: number;
  private readonly sender: Sender;

  private timer: ReturnType<typeof setInterval> | null = null;
  private flushing = false;
  private stopped = false;
  private _dropped = 0;

  // Tracks in-flight + buffered work so flush()/shutdown() can wait for drain.
  private inFlight = 0;

  constructor(baseUrl: string, apiKey: string, opts: TelemetryOptions = {}) {
    this.maxQueue = opts.maxQueue ?? 10_000;
    this.batchSize = opts.batchSize ?? 100;
    this.flushIntervalMs = opts.flushIntervalMs ?? 1000;
    this.maxRetries = opts.maxRetries ?? 3;
    this.sender =
      opts.sender ?? new HttpSender(baseUrl, apiKey, opts.timeoutMs ?? 5000);

    this.timer = setInterval(() => {
      void this.tick();
    }, this.flushIntervalMs);
    // Don't keep the event loop alive solely for telemetry.
    if (typeof this.timer.unref === "function") this.timer.unref();
  }

  get dropped(): number {
    return this._dropped;
  }

  /** Non-blocking enqueue. Drops + warns if the buffer is full. */
  enqueue(event: Event | Record<string, unknown>): void {
    if (this.stopped) return;
    if (this.buffer.length >= this.maxQueue) {
      this._dropped += 1;
      warn(
        `telemetry buffer full (max=${this.maxQueue}), dropping event (total dropped=${this._dropped})`,
      );
      return;
    }
    this.buffer.push(event as Record<string, unknown>);
    // Eagerly flush when a full batch has accumulated (batch-by-N).
    if (this.buffer.length >= this.batchSize) {
      void this.tick();
    }
  }

  /** Drain whole batches from the buffer and ship them. Re-entrancy guarded. */
  private async tick(): Promise<void> {
    if (this.flushing) return;
    this.flushing = true;
    try {
      while (this.buffer.length > 0) {
        const batch = this.buffer.splice(0, this.batchSize);
        this.inFlight += 1;
        try {
          await this.shipWithRetry(batch);
        } finally {
          this.inFlight -= 1;
        }
      }
    } finally {
      this.flushing = false;
    }
  }

  private async shipWithRetry(batch: Array<Record<string, unknown>>): Promise<void> {
    if (batch.length === 0) return;
    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      let status: number | null;
      try {
        status = await this.sender.send(batch);
      } catch (e) {
        // Sender should not throw, but degrade gracefully if it does.
        warn(
          `ingest send failed (attempt ${attempt + 1}/${this.maxRetries}): ${String(e)}`,
        );
        status = null;
      }

      if (status === 202) return;
      if (status === 401) {
        warn(`ingest auth failed (401), dropping ${batch.length} events`);
        return;
      }
      if (status === 429) {
        // Backpressure — back off and retry this same batch.
        const backoff = Math.min(2 ** attempt * 250, 5000);
        await sleep(backoff);
        continue;
      }
      if (status === null) {
        const backoff = Math.min(2 ** attempt * 100, 2000);
        await sleep(backoff);
        continue;
      }
      // Other non-retryable status.
      warn(`ingest returned status ${status}, dropping ${batch.length} events`);
      return;
    }

    this._dropped += batch.length;
    warn(
      `dropping ${batch.length} events after ${this.maxRetries} failed attempts (total dropped=${this._dropped})`,
    );
  }

  /** Block until the buffer is empty and in-flight sends settle (best effort). */
  async flush(timeoutMs = 5000): Promise<boolean> {
    const deadline = Date.now() + timeoutMs;
    void this.tick();
    while (Date.now() < deadline) {
      if (this.buffer.length === 0 && this.inFlight === 0 && !this.flushing) {
        return true;
      }
      await sleep(10);
      if (this.buffer.length > 0 && !this.flushing) void this.tick();
    }
    return this.buffer.length === 0 && this.inFlight === 0;
  }

  /** Drain and stop. Idempotent. */
  async shutdown(timeoutMs = 5000): Promise<void> {
    if (this.stopped) return;
    this.stopped = true;
    if (this.timer !== null) {
      clearInterval(this.timer);
      this.timer = null;
    }
    await this.flush(timeoutMs);
    if (this.sender.close) {
      try {
        await this.sender.close();
      } catch {
        // ignore
      }
    }
  }
}

File: sdks/python/costobs/telemetry.py
"""Background telemetry shipping.

Design goals (from the spec):
  * The synchronous hot path is a single ``queue.put_nowait``.
  * A daemon thread batches events by count (N) or time (T) and POSTs them.
  * Graceful degradation: unreachable ingest -> log + drop (after retries).
    A 429 backpressure response is retried with backoff, not dropped.
  * Bounded queue: when full, ``put_nowait`` raises ``Full`` -> drop + warn,
    never block the caller.
"""

from __future__ import annotations

import atexit
import json
import logging
import queue
import threading
import time
from typing import Any, Dict, List, Optional

logger = logging.getLogger("costobs")

# Prefer httpx; fall back to urllib so the SDK works with stdlib only.
try:  # pragma: no cover - import shim
    import httpx  # type: ignore

    _HAS_HTTPX = True
except Exception:  # pragma: no cover
    httpx = None  # type: ignore
    _HAS_HTTPX = False

_SENTINEL = object()


class _HttpSender:
    """Thin POST wrapper. Returns the HTTP status code, or None on transport error."""

    def __init__(self, base_url: str, api_key: str, timeout: float = 5.0):
        self.url = base_url.rstrip("/") + "/v1/events"
        self.api_key = api_key
        self.timeout = timeout
        self._client = httpx.Client(timeout=timeout) if _HAS_HTTPX else None

    def send(self, events: List[Dict[str, Any]]) -> Optional[int]:
        body = json.dumps({"events": events}).encode("utf-8")
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }
        if _HAS_HTTPX:
            resp = self._client.post(self.url, content=body, headers=headers)
            return resp.status_code
        import urllib.error
        import urllib.request

        req = urllib.request.Request(self.url, data=body, headers=headers, method="POST")
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                return resp.getcode()
        except urllib.error.HTTPError as e:
            return e.code

    def close(self) -> None:
        if self._client is not None:
            try:
                self._client.close()
            except Exception:
                pass


class TelemetryQueue:
    """Bounded queue + daemon sender thread."""

    def __init__(
        self,
        base_url: str,
        api_key: str,
        *,
        max_queue: int = 10_000,
        batch_size: int = 100,
        flush_interval: float = 1.0,
        max_retries: int = 3,
        sender: Optional[Any] = None,
        timeout: float = 5.0,
    ):
        self._q: "queue.Queue[Any]" = queue.Queue(maxsize=max_queue)
        self._batch_size = batch_size
        self._flush_interval = flush_interval
        self._max_retries = max_retries
        self._sender = sender if sender is not None else _HttpSender(base_url, api_key, timeout)
        self._stop = threading.Event()
        self._drained = threading.Event()
        self._dropped = 0
        self._dropped_lock = threading.Lock()
        self._thread = threading.Thread(
            target=self._run, name="costobs-telemetry", daemon=True
        )
        self._thread.start()
        atexit.register(self.shutdown)

    # ---- producer side (hot path) ----

    def enqueue(self, event_dict: Dict[str, Any]) -> None:
        """Non-blocking enqueue. Drops + warns if the queue is full."""
        try:
            self._q.put_nowait(event_dict)
        except queue.Full:
            with self._dropped_lock:
                self._dropped += 1
                dropped = self._dropped
            logger.warning(
                "costobs: telemetry queue full (max=%d), dropping event (total dropped=%d)",
                self._q.maxsize,
                dropped,
            )

    @property
    def dropped(self) -> int:
        with self._dropped_lock:
            return self._dropped

    # ---- consumer side (background thread) ----

    def _run(self) -> None:
        batch: List[Dict[str, Any]] = []
        deadline = time.monotonic() + self._flush_interval
        while True:
            timeout = max(0.0, deadline - time.monotonic())
            try:
                item = self._q.get(timeout=timeout)
            except queue.Empty:
                item = None

            if item is _SENTINEL:
                # Drain everything still queued, then flush and exit.
                self._flush(batch)
                self._mark_done(len(batch))
                batch = []
                self._q.task_done()  # the sentinel itself
                self._drain_remaining()
                self._drained.set()
                return

            if item is not None:
                batch.append(item)

            now = time.monotonic()
            if len(batch) >= self._batch_size or (batch and now >= deadline):
                self._flush(batch)
                self._mark_done(len(batch))
                batch = []
                deadline = now + self._flush_interval
            elif not batch:
                deadline = now + self._flush_interval

    def _mark_done(self, n: int) -> None:
        """Account for n consumed items so ``Queue.unfinished_tasks`` (and thus
        :meth:`flush`) reflects fully-shipped work, not just dequeued work."""
        for _ in range(n):
            self._q.task_done()

    def _drain_remaining(self) -> None:
        leftovers: List[Dict[str, Any]] = []
        while True:
            try:
                item = self._q.get_nowait()
            except queue.Empty:
                break
            if item is _SENTINEL:
                self._q.task_done()
                continue
            leftovers.append(item)
            if len(leftovers) >= self._batch_size:
                self._flush(leftovers)
                self._mark_done(len(leftovers))
                leftovers = []
        if leftovers:
            self._flush(leftovers)
            self._mark_done(len(leftovers))

    def _flush(self, batch: List[Dict[str, Any]]) -> None:
        if not batch:
            return
        for attempt in range(self._max_retries):
            try:
                status = self._sender.send(batch)
            except Exception as e:  # transport failure
                logger.warning(
                    "costobs: ingest send failed (attempt %d/%d): %s",
                    attempt + 1,
                    self._max_retries,
                    e,
                )
                status = None

            if status == 202:
                return
            if status == 401:
                logger.warning("costobs: ingest auth failed (401), dropping %d events", len(batch))
                return
            if status == 429:
                # Backpressure — back off and retry this same batch.
                backoff = min(2.0 ** attempt * 0.25, 5.0)
                logger.debug("costobs: ingest backpressure (429), retrying in %.2fs", backoff)
                time.sleep(backoff)
                continue
            if status is None:
                backoff = min(2.0 ** attempt * 0.1, 2.0)
                time.sleep(backoff)
                continue
            # Other non-retryable status.
            logger.warning("costobs: ingest returned status %s, dropping %d events", status, len(batch))
            return

        with self._dropped_lock:
            self._dropped += len(batch)
            dropped = self._dropped
        logger.warning(
            "costobs: dropping %d events after %d failed attempts (total dropped=%d)",
            len(batch),
            self._max_retries,
            dropped,
        )

    # ---- lifecycle ----

    def flush(self, timeout: float = 5.0) -> bool:
        """Block until every enqueued event has been shipped (or dropped).

        ``unfinished_tasks`` counts puts minus ``task_done`` calls; the sender
        thread only marks items done after the batch containing them has been
        flushed, so reaching zero means nothing is queued *or* in flight.
        Returns True if fully drained within ``timeout``.
        """
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            if self._q.unfinished_tasks == 0:
                return True
            time.sleep(0.01)
        return self._q.unfinished_tasks == 0

    def shutdown(self, timeout: float = 5.0) -> None:
        """Signal the sender to drain and stop. Idempotent."""
        if self._stop.is_set():
            return
        self._stop.set()
        try:
            self._q.put_nowait(_SENTINEL)
        except queue.Full:
            # Make room — drop one event so the sentinel lands.
            try:
                self._q.get_nowait()
                self._q.task_done()  # account for the dropped event
                self._q.put_nowait(_SENTINEL)
            except Exception:
                pass
        self._drained.wait(timeout=timeout)
        if hasattr(self._sender, "close"):
            self._sender.close()
